Supporting multi-chain mcmc with cuda tensors

fehiepsi opened the pull request:
This PR is just used for discussion with @neerajprad .

Hi Neeraj, I added events to workers in ParallelSampler so we can solve the problem at https://discuss.pytorch.org/t/cuda-tensors-on-multiprocessing-queue/28626.

However, there is also another problem: [Invalid device pointer](https://discuss.pytorch.org/t/invalid-device-pointer-using-multiprocessing-with-cuda/28023).

After debugging, I noticed that if we clone `args` and `kwargs` in the returned `trace` of each mcmc step, then this error will go away. Of course, if we only put `z` (instead of trace) to the Queue (by returning `z` at the `sample()` method of HMC/NUTS), and get_trace from z retrieved from Queue, things will be fine (_note that because we cache (z, pe, grad), we don't need to use previous trace for the next sampling step_). However, it will add overhead to the workers: paused workers have to wait for get_trace from z of the current worker.

Maybe I am missing something so I make PR to discuss with you about solutions. :)

neerajprad commented:
@fehiepsi - Thanks for exploring this! I think one issue here is that if we use `mp.Event` to signal from the main thread that the corresponding tensor from the worker process has been read, the worker processes might spend a lot of time waiting, reducing the advantage of parallelizing over separate processes. What is the time advantage that you notice between `num_chains=1` and `num_chains=4` in your current solution? 

I will play around with your PR a bit. I think a faster solution may be to hold references to the tensors in the worker threads (say in a list), and clear them off based on some periodic communication from the main thread. 

Regarding, 

> I noticed that if we clone args and kwargs in the returned trace of each mcmc step, then this error will go away.

I think we need to only do this once when we initialize the worker, i.e. if you clone in `init_workers`, it will probably work fine, and be a much simpler change. The reason is that the trace contains certain tensors that are being passed back and forth between the main process and the worker process (see https://github.com/pytorch/pytorch/issues/10375 for details) which doesn't jive well with cuda multiprocessing. I think I had some branch with those changes, let me see if I can update and push that upstream.

fehiepsi commented:
@neerajprad I made a benchmark [here](https://gist.github.com/fehiepsi/3b4cb73b34997be01fd0842d147c7263) FYI. For slow model, the overhead is negligible as expected. For very fast model, the overhead is noticeable but it is a small trade-off if we want to use this approach to support multi-chain in GPU.

neerajprad commented:
@fehiepsi - The performance seems quite reasonable. This will also help fix https://github.com/uber/pyro/issues/1699 (exc. for the `pyro.param` issue, see below).

Could you update this PR to do the following:
 - Remove the catch-pass hack `except socket.error as e` in `_ParallelSampler._traces`. We shouldn't hit this bug now given that the workers are alive until the tensors are read from the main thread.
 - Add `--cuda` option to `baseball.py` example.
 - Enable the diagnostics test in `test_mcmc` with `num_chains=2` for testing locally, and remove `pyro.param` statement from the model. The global param store does not play well with multiprocessing, and we can debug this later.

fehiepsi commented:
@neerajprad Will we address the error **Invalid device pointer.** later? The changes in this PR have not solve that problem for CUDA tensors yet. And I don't have further ideas how to solve it. :(

neerajprad commented:
> Will we address the error Invalid device pointer. 

Sorry, I forgot about that issue. I will try to push a fix for that in this PR itself.

fehiepsi commented:
@neerajprad I have addressed your suggestions. I can confirmed that this fixes the issue at #1699. The only issue now is to make this work for CUDA tensors. I let it for you to make any change you want. :)

neerajprad commented:
@fehiepsi - I am seeing a different issue now. We can fix the `invalid device pointer` by cloning the args (assume for the moment that all tensors are passed in args as we do for JitTraceELBO). Even if we consider models which don't take in any args/kwargs as inputs, we see HMC fail with:

```sh
  File "/home/npradhan/miniconda3/envs/pytorch/lib/python3.6/site-packages/torch/multiprocessing/reductions.py", line 104, in rebuild_cuda_tensor
    t = torch._utils._rebuild_tensor(storage, tensor_offset, tensor_size, tensor_stride)
  File "/home/npradhan/miniconda3/envs/pytorch/lib/python3.6/site-packages/torch/_utils.py", line 131, in _rebuild_tensor
    return tensor_class().set_(storage, storage_offset, size, stride)
RuntimeError: Assertion `self->allocator() != nullptr' failed.  at /pytorch/aten/src/THC/THCStorage.cpp:16
``` 

Earlier, I wasn't seeing this issue. I was getting garbage values instead, without any exceptions being thrown. I am not sure if this is due to some change in PyTorch 1.0, or some recent changes within the `mcmc` module.

fehiepsi commented:
Do you get the same error for PriorKernel diagnostics test (with mp_context="spawn")? Before 1.0 release, there are some changes in pytorch multiprocessing module (e.g. [this pull request](https://github.com/pytorch/pytorch/pull/14736)). I don't think that changes in NUTS/HMC are relevant.

neerajprad commented:
> Do you get the same error for PriorKernel diagnostics test (with mp_context="spawn")? 

No, this is only an issue with the NUTS / HMC kernel, and I wonder if it is related to caching the previous trace.

fehiepsi commented:
> I wonder if it is related to caching the previous trace.

I think we can detach and clone them to see if that is the problem. I can't make this work for PriorKernel so I can't go further to test. :(

fehiepsi commented:
@neerajprad I can confirm that your fixes solve the "Invalid device pointer" problem. For the error: _RuntimeError: Assertion `self->allocator() != nullptr' failed._, the following script (which is not related to MCMC) replicates it
```
import torch
import torch.multiprocessing as mp

import pyro
import pyro.distributions as dist

torch.set_default_tensor_type(torch.cuda.FloatTensor)
n = 10

def model():
    loc = pyro.sample("loc", dist.Normal(torch.zeros(3), 1))
    pyro.sample("y", dist.Normal(loc, 1))  # comment out this line -> no error

def worker(q, e):
    for i in range(n):
        trace = {"normal": dist.Normal(0, 1)}  # comment out this line -> no error ???
        trace = pyro.poutine.trace(model).get_trace()
        q.put(trace.nodes["loc"])
        e.wait()
        e.clear()

if __name__ == "__main__":
    ctx = mp.get_context("spawn")
    q = ctx.Queue()
    e = ctx.Event()
    p = ctx.Process(target=worker, args=(q, e))
    p.start()
    for i in range(n):
        trace = q.get()
        e.set()
    p.join()
```
I am not sure if this is a PyTorch issue or Pyro issue. I think that it is PyTorch issue but I can't make an example without using `pyro.poutine.trace`. When I remove some lines in the above example (where I put the comments), no error happened, which is mysterious to me.

One thing I notice during debugging is when we put `trace` to the Queue, there is a mismatch of the storage sizes. [This line](https://github.com/pytorch/pytorch/blob/master/torch/multiprocessing/reductions.py#L94) returns a storage with size **less than** `storage_size_bytes`, after that the error happened.

While playing with PriorKernel (using data with shape 3, instead of 1), I notice that in some iteration, we have a storage with size **larger than** `storage_size_bytes`. This will not throw the above error (because to reconstruct the desired tensor, we just need to get the first `storage_size_bytes` of that storage) but I think that it will also make things wrong for a more complicated kernel.

This is a really complicated issue to debug (took me a whole day to come up with the above example), so I make this low priority. We can play around the above example for fun to get to the real issue if we have more time. :)

neerajprad commented:
Thanks for the example here, I will play around with it. Do you still see an error if you only put the tensor on to the queue, `q.put(trace.nodes["loc"]["value"])` (I'm guessing not)?

fehiepsi commented:
Yes, it does not. I'm using pytorch 1.0.0.dev20181202 and am able to run MCMC sampler with multi-chain. However, execute_traces seems to hold wrong values. I am investigating the reason.

fehiepsi commented:
@neerajprad The PR https://github.com/pytorch/pytorch/pull/16194 fixed **Assertion self->allocator() != nullptr failed** issue. However, we have to clone the traces after retrieving because of the reason at https://github.com/pytorch/pytorch/issues/16141#issuecomment-455925568. But let's address that issue in another PR if needed.

neerajprad commented on pyro/infer/mcmc/mcmc.py:
```diff
@@ -87,6 +88,8 @@ def run(self, *args, **kwargs):
         try:
             for sample in self.trace_gen._traces(*args, **kwargs):
                 self.result_queue.put_nowait((self.chain_id, sample))
+                self.event.wait()
```
Can we evaluate the effect of adding these lines for the CPU multiprocessing case? Does this put any overhead on existing models (both already slow or fast)? I suppose the answer has to be yes to that, but I am curious how much of an overhead it is.

neerajprad commented:
Thanks for following up on this!

> However, we have to clone the traces after retrieving because of the reason at pytorch/pytorch#16141 (comment). 

This seems like a major deficiency to me in that I am not sure if this will scale for larger data (precisely the kinds where we might expect the GPU to be actually useful), and seems to defeat the point of using shared memory.

neerajprad commented on pyro/infer/mcmc/hmc.py:
```diff
@@ -408,7 +408,8 @@ def sample(self, trace):
             accept_prob = (-delta_energy).exp().clamp(max=1.)
-        rand = pyro.sample("rand_t={}".format(self._t), dist.Uniform(torch.zeros(1), torch.ones(1)))
+        rand = pyro.sample("rand_t={}".format(self._t), dist.Uniform(accept_prob.new_tensor(0.),
+                                                                     accept_prob.new_tensor(1.)))
```
+1 on this..I have been wanting to make this change for some time!

fehiepsi commented:
Yes, how about not using shared memory? Could you point out some advantages of using shared memory?

fehiepsi commented:
I see. Let me try to hold all the traces in the workers and move them to the main process before termination. Hope that work.

fehiepsi commented:
Thanks! It makes sense. Making sure that all the changes won't affect the current performance is top priority.

In case we want to draw a huge amount of samples, we have to think about memory issue. What do you think about this sketch:
+ Though workers are generators, they also keep traces in a list.
+ When we *reduce them to some intermediate / terminal representation*, we send a signal back to the workers to destroy the list?

neerajprad commented:
> When we reduce them to some intermediate / terminal representation, we send a signal back to the workers to destroy the list?

That seems like a reasonable solution, but we can defer that discussion to a future PR. I think this PR is fine as is (it has a bunch of fixes and doesn't any major design changes that require discussion) provided that we aren't making the CPU runtime worse with the *process wait / resume*. We might need to confirm that using different models, both fast and slow. I am happy to merge once you confirm that perf on the CPU remains unimpacted.

fehiepsi commented:
:D it has been a long time for us to discuss on this so you might forget that we have done a benchmark. I did a benchmark at https://gist.github.com/fehiepsi/3b4cb73b34997be01fd0842d147c7263, which tested both fast and slow models. For slow models, times are equivalent (sometime this PR is faster, sometime it is slower, depending on the system situation when running; but they are just different by a very small margin). For fast models, it is noticeable. With my system, this PR will be slower by 1s if we draw 4000 samples.
